btree/btree.c

```diff
@@ -3529,21 +3559,429 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
+ * other child and accordingly return the next state for execution.
+ *
+ * @param bop It will provide all required information about btree operation.
+ * @return int64_t it return state which needs to get executed next.
+ */
+static int64_t handle_root_case(struct m0_btree_op *bop)
+{
+	/**
+	 * If root is internal node and it contains only two records, check if
+	 * any record is going to be deleted if yes, we also have to load other
+	 * child of root so that we can copy the content from that child at root
+	 * and decrease the level by one.
+	 */
+	struct m0_btree_oimpl *oi = bop->bo_i;
+	int8_t                 load;
+
+	load = root_child_is_req(bop);
+	if (load == -1) {
+		level_cleanup(oi, bop->bo_tx);
+		return P_SETUP;
+	}
+	if (load) {
+		struct slot    root_slot = {};
+		struct segaddr root_child;
+		struct level   *root_lev = &oi->i_level[0];
+
+		root_slot.s_node = root_lev->l_node;
+		root_slot.s_idx  = root_lev->l_idx == 0 ? 1 : 0;
+
+		node_child(&root_slot, &root_child);
+		if (!address_in_segment(root_child)) {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, M0_ERR(-EFAULT));
+		}
+
+		return node_get(&oi->i_nop, bop->bo_arbor->t_desc,
+				&root_child, P_STORE_CHILD);
+	}
+	return P_LOCK;
+}
+
+/* State machine implementation for delete operation */
+static int64_t btree_del_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop        = M0_AMB(bop, smop, bo_op);
+	struct td             *tree       = bop->bo_arbor->t_desc;
+	uint64_t               flags      = bop->bo_flags;
+	struct m0_btree_oimpl *oi         = bop->bo_i;
+	struct level          *lev;
+
+	if(oi)
+		lev = &oi->i_level[oi->i_used];
+
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if ((flags & BOF_COOKIE) &&
+		    cookie_is_set(&bop->bo_rec.r_key.k_cookie))
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE:{
+		void *cookie_segaddr = bop->bo_rec.r_key.k_cookie.segaddr;
+		if (cookie_is_valid(tree, &bop->bo_rec.r_key.k_cookie) &&
+		    !cookie_underflow_is_possible(cookie_segaddr))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	}
+	case P_SETUP: {
+		bop->bo_arbor->t_height = tree->t_height;
+		bop->bo_i = level_alloc(bop->bo_arbor->t_height);
+		if (bop->bo_i == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		bop->bo_i->i_key_found = false;
+		return P_LOCKALL;
+	}
+	case P_LOCKALL:
+		if (bop->bo_flags & BOF_LOCKALL) {
+			//return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+			m0_rwlock_write_lock(&tree->t_lock);
+		}
+		/* Fall through to the next stage */
+	case P_DOWN:
+		oi->i_used = 0;
+		/* Load root node. */
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr,
+				P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    node_slot = {};
+			struct segaddr child_node_addr;
+
+			lev->l_node = oi->i_nop.no_node;
+			node_slot.s_node = oi->i_nop.no_node;
+			lev->l_seq = lev->l_node->n_seq;
+			/* verify node footer */
+			/*int rc = node_verify(lev->l_node);
```

Comment:
as node_verify will give wrong output as we have not set footer for node so it will return wrong rc.

Reply:
Can you for now uncomment the node_verify function from the code and return TRUE. Also add a comment to work on that function in future to return the correct output. Also I think the return value should be bool instead of int.

Reply:
What is the point of returning an error code because the caller cannot do anything specific based on the error code which this routine returns. The caller just needs to know if the node is valid and the verify needs to return only true/false.